redis/cluster.py

```diff
@@ -0,0 +1,1884 @@
+                           server_type=None):
+        """
+        Gets a node that servers this hash slot
+        """
+        if self._moved_exception:
+            with self._lock:
+                if self._moved_exception:
+                    self._update_moved_slots()
+
+        if self.slots_cache.get(slot) is None or \
+                len(self.slots_cache[slot]) == 0:
+            raise SlotNotCoveredError(
+                'Slot "{0}" not covered by the cluster. '
+                '"require_full_coverage={1}"'.format(
+                    slot, self._require_full_coverage)
+            )
+
+        if read_from_replicas:
+            # get the server index in a Round-Robin manner
+            primary_name = self.slots_cache[slot][0].name
+            node_idx = self.read_load_balancer.get_server_index(
+                primary_name, len(self.slots_cache[slot]))
+        elif (
+                server_type is None
+                or server_type == PRIMARY
+                or len(self.slots_cache[slot]) == 1
+        ):
+            # return a primary
+            node_idx = 0
+        else:
+            # return a replica
+            # randomly choose one of the replicas
+            node_idx = random.randint(
+                1, len(self.slots_cache[slot]) - 1)
+
+        return self.slots_cache[slot][node_idx]
+
+    def get_nodes_by_server_type(self, server_type):
+        return [
+            node
+            for node in self.nodes_cache.values()
+            if node.server_type == server_type
+        ]
+
+    def populate_startup_nodes(self, nodes):
+        """
+        Populate all startup nodes and filters out any duplicates
+        """
+        for n in nodes:
+            self.startup_nodes[n.name] = n
+
+    def cluster_require_full_coverage(self, cluster_nodes):
+        """
+        if exists 'cluster-require-full-coverage no' config on redis servers,
+        then even all slots are not covered, cluster still will be able to
+        respond
+        """
+
+        def node_require_full_coverage(node):
+            try:
+                return ("yes" in node.redis_connection.config_get(
+                    "cluster-require-full-coverage").values()
+                )
+            except ConnectionError:
+                return False
+            except Exception as e:
+                raise RedisClusterException(
+                    'ERROR sending "config get cluster-require-full-coverage"'
+                    ' command to redis server: {0}, {1}'.format(node.name, e)
+                )
+
+        # at least one node should have cluster-require-full-coverage yes
+        return any(node_require_full_coverage(node)
+                   for node in cluster_nodes.values())
+
+    def check_slots_coverage(self, slots_cache):
+        # Validate if all slots are covered or if we should try next
+        # startup node
+        for i in range(0, REDIS_CLUSTER_HASH_SLOTS):
```

Comment:
```
if slots_cache not in range(0, REDIS_CLUSTER_HASH_SLOTS):
    return False
return True
```

Reply:
It will not give us the required result. We want to check if all slots from 0 to REDIS_CLUSTER_HASH_SLOTS are found in slots_cache, what you suggests mean that if we'll have slots not in the above range we'll get false, but we can have only one slot in slots_cache and it will return true...